### PR TITLE
Rust replacements for cc tests

### PR DESCRIPTION
**Pull request by newpavlov**
#203 
I agree to license my contributions to each file under the terms given at the top of each file I changed.

### CODE REVIEW

**Comment by newpavlov** [comment]
Merged with samscott89 code to make it easier to merge it. Now this PR covers also #202.

**Comment by briansmith** [comment]
Thanks a lot for this. And, thanks for rebasing on @samscott89's stuff. I'm working on the backlog of PRs today, and I really appreciate that.

Please add the "I agree to license..." to the bottom of each commit (not the bottom of the PR message). Thanks!

It will likely be a few days at least before I can merge this, because of the backlog. I hope to be more responsive in the future.

**Comment by newpavlov** [comment]
Added `aes_test` for #204.
One thing: I am not too confident about my FFI code (don't have much experince with it) so it's better to check it more closely.

**Comment by briansmith** [comment]
This is all awesome work. Two things:

1. I have something I need to finish today, which I need to monopolize the Travis CI for. Don't get discouraged if I cancel your builds. I will restart them when I am done with the more urgent stuff.

2. Please split this up into multiple PRs, one for Poly1305 and one for AES. Note that when you create AES PR, or after you create it, you can set the base branch of the AES PR to be the branch you used for the poly1305 PR. Splitting this up into multple PRs makes it easier for me to review incrementally and is much appreciated.

Thanks again!

**Comment by newpavlov** [comment]
About `aes_test` #204. It will require a bit of handwork to decouple it properly into separate PR as I worked with master branches and unfortunately haven't created a separate branch based on upstream for aes_test commit. Also I only able to choose base branch fr PR only from your repo not mine.

I will tend to it later if PR in this from will not be merged. Luckily there is not many changes in this PR, so I hope you will be able to review it in the current form.

**Comment by briansmith** [comment]
OK, no problem.

**Comment by briansmith** [comment]
Because I landed the ChaCha20 test rewrite already, and because of various refactorings and renamings that had to be done urgently, this PR is now bitrotten. But, I updated it and the updated form is now at in the [oxidize-poly1305-tests](https://github.com/briansmith/ring/tree/oxidize-poly1305-tests) branch. I'll work on reviewing the actual changes soon.

File: src/aead/aes_gcm.rs [hunk 1]
```diff
@@ -58,11 +58,9 @@ pub static AES_256_GCM: aead::Algorithm = aead::Algorithm {
     open: aes_gcm_open,
 };
 
-fn aes_gcm_init(ctx_buf: &mut [u8], key: &[u8])
-                -> Result<(), error::Unspecified> {
+fn aes_gcm_init(ctx_buf: &mut [u8], key: &[u8]) -> Result<(), error::Unspecified> {
     bssl::map_result(unsafe {
-        evp_aead_aes_gcm_init(ctx_buf.as_mut_ptr(),
-                              ctx_buf.len(), key.as_ptr(), key.len())
+        evp_aead_aes_gcm_init(ctx_buf.as_mut_ptr(), ctx_buf.len(), key.as_ptr(), key.len())
```
Comment (briansmith) [inline src/aead/aes_gcm.rs]: Is this just the result of running rustfmt? Let's keep the changes in this PR specifically to the tests, so let's remove this.

File: src/aead/aes_gcm.rs [hunk 2]
```diff
@@ -129,4 +127,56 @@ mod tests {
...
+        buf
+    }
+
+    #[test]
+    pub fn test_aes() {
+        test::from_file("src/aead/aes_tests.txt", |section, test_case| {
+            assert_eq!(section, "");
+            let key = test_case.consume_bytes("Key");
+            let input = convert_block(&test_case.consume_bytes("Input"));
+            let output = convert_block(&test_case.consume_bytes("Output"));
```
Comment (briansmith) [inline src/aead/aes_gcm.rs]: Instead of doing this, let's do something like this:

```
let input = test_case.consume_bytes("Input");
let input = slice_as_array_ref!(&input, AES_BLOCK_SIZE);
let output = test_case.consume_bytes("Output");
let output = slice_as_array_ref!(&output, AES_BLOCK_SIZE);
```

File: src/aead/aes_gcm.rs [hunk 3]
```diff
@@ -129,4 +127,56 @@ mod tests {
...
+                assert_eq!(buf, output);
+            };
+            Ok(())
+        })
+    }
+
+    const AES_BLOCK_SIZE: usize = 16;
+
+    #[repr(C)]
+    pub struct AesKey {
```
Comment (briansmith) [inline src/aead/aes_gcm.rs]: The conventions in *ring* are:

* Use the same name as the C type. In this case, `AES_KEY`. You'd need to do `#[allow(non_camel_case_types)]`.
* Add a comment "// Keep this in sync with `AES_KEY` in aes.h"
* Add a similar comment above AES_KEY in aes.h saying "/* Keep this in sync with AES_KEY in aes_gcm.rs. */

File: src/aead/aes_gcm.rs [hunk 4]
```diff
@@ -129,4 +127,56 @@ mod tests {
...
+            let mut aes_key = AesKey {
+                rd_key: [0u32; 4 * (AES_MAX_ROUNDS + 1)],
+                rounds: 0,
+            };
+
+            unsafe {
+                // Key setup
+                if AES_set_encrypt_key(key.as_ptr(), key.len() * 8, &mut aes_key) != 0 {
+                    panic!("Failed to set aes key");
+                }
```
Comment (briansmith) [inline src/aead/aes_gcm.rs]: You can just do `assert_eq!(GFp_AES_set_encrypt_key(...), 0, "GFp_AES_set_encrypt_key failed");`

File: src/aead/aes_gcm.rs (hunk 2 above)
Comment (briansmith) [inline src/aead/aes_gcm.rs]: I suggest you rename `output` as `expected_output`.

File: src/aead/aes_gcm.rs [hunk 5]
```diff
@@ -129,4 +127,56 @@ mod tests {
...
+
+                // Encryption into separate buffer
+                let mut buf = [0u8; AES_BLOCK_SIZE];
+                AES_encrypt(input.as_ptr(), buf.as_mut_ptr(), &aes_key);
+                assert_eq!(buf, output);
```
Comment (briansmith) [inline src/aead/aes_gcm.rs]: I suggest you rename `buf` to `actual_output`.

File: src/aead/aes_gcm.rs [hunk 6]
```diff
@@ -129,4 +127,56 @@ mod tests {
...
+
+                // In-place encryption
+                buf.clone_from_slice(&input[..]);
```
Comment (briansmith) [inline src/aead/aes_gcm.rs]: I think we can use `copy_from_slice` instead?

File: src/aead/chacha20_poly1305.rs [hunk 7]
```diff
@@ -50,14 +50,12 @@ fn chacha20_poly1305_open(ctx: &[u64; aead::KEY_CTX_BUF_ELEMS],
...
-fn chacha20_poly1305_update(state: &mut [u8; POLY1305_STATE_LEN],
-                            ad: &[u8], ciphertext: &[u8]) {
+fn chacha20_poly1305_update(state: &mut [u8; POLY1305_STATE_LEN], ad: &[u8], ciphertext: &[u8]) {
     fn update_padded_16(state: &mut [u8; POLY1305_STATE_LEN], data: &[u8]) {
         poly1305_update(state, data);
         if data.len() % 16 != 0 {
             static PADDING: [u8; 16] = [0u8; 16];
-            poly1305_update(state,
-                            &PADDING[..PADDING.len() - (data.len() % 16)])
+            poly1305_update(state, &PADDING[..PADDING.len() - (data.len() % 16)])
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: I guess this is due to rustfmt? We should remove these non-test changes and/or move them to another PR.

File: src/aead/chacha20_poly1305.rs [hunk 8]
```diff
@@ -256,29 +244,111 @@ extern {
...
 
     #[test]
     pub fn test_poly1305_state_len() {
-        assert_eq!((super::POLY1305_STATE_LEN + 255) / 256,
-                    (CRYPTO_POLY1305_STATE_LEN + 255) / 256);
+        assert_eq!((POLY1305_STATE_LEN + 255) / 256,
+                   (CRYPTO_POLY1305_STATE_LEN + 255) / 256);
+    }
+
+    fn test_simd(excess: usize, key: [u8; POLY1305_KEY_LEN],
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: Because this is significantly transliterated from Google's code, we need to add a line "// Portions Copyright (c) 2015, Google Inc.", verbatim, underneath the copyright line at the top of this file.

File: src/aead/chacha20_poly1305.rs [hunk 9]
```diff
@@ -256,29 +244,111 @@ extern {
...
+        let mut mac_out = [0u8; aead::TAG_LEN];
+        poly1305_init(&mut state, &key);
+
+        // Feed 16 bytes in. Some implementations begin in non-SIMD mode and
+        // upgrade on-demand. Stress the upgrade path.
+        let init = if input.len() < 16 { input.len() } else { 16 };
+
+        poly1305_update(&mut state, &input[..init]);
+        for chunk in input[init..].chunks(128 + 2 * excess) {
+            let (long, short) = if chunk.len() < (128 + excess) {
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: I don't think we should have the parenthese around (128 + excess).

File: src/aead/chacha20_poly1305.rs (hunk 8 above)
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: The convention in *ring* is that we put helper functions like `test_simd` either nested inside, or underneath, the function they help (test_poly1305 in this case). Please move this below that function.

File: src/aead/chacha20_poly1305.rs [hunk 10]
```diff
@@ -256,29 +244,111 @@ extern {
...
+                chunk.split_at(128 + excess)
+            };
+            // Feed 128 + |excess| bytes to test SIMD mode
+            poly1305_update(&mut state, long);
+            // Feed |excess| bytes to ensure SIMD mode can handle short inputs
+            if !short.is_empty() {
+                poly1305_update(&mut state, short);
+            }
+        }
+        poly1305_finish(&mut state, &mut mac_out);
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: Let's move the declaration/initialization of `mac_out` here. I suggest renaming `mac` to `expected_mac` and rename `mac_out` to `actual_mac`.

File: src/aead/chacha20_poly1305.rs [hunk 11]
```diff
@@ -256,29 +244,111 @@ extern {
...
+        if mac != mac_out {
+            println!("SIMD pattern {} failed.", excess);
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: Let's just use `assert_eq!()` here. Avoiding explicit `println!` will make it easier to port the tests to `#![no_std]` platforms in the future.

File: src/aead/chacha20_poly1305.rs [hunk 12]
```diff
@@ -256,29 +244,111 @@ extern {
...
+        test::from_file("src/aead/poly1305_test.txt", |section, test_case| {
+            assert_eq!(section, "");
+            let key_bytes = test_case.consume_bytes("Key");
+            let input = test_case.consume_bytes("Input");
+            let mac_bytes = test_case.consume_bytes("MAC");
+
+            let key = {
+                assert_eq!(key_bytes.len(), POLY1305_KEY_LEN);
+                let mut buf = [0u8; POLY1305_KEY_LEN];
+                buf.clone_from_slice(&key_bytes);
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: NIT: `copy_from_slice` (and below).

File: src/aead/chacha20_poly1305.rs [hunk 13]
```diff
@@ -256,29 +244,111 @@ extern {
...
+            };
+
+            let mac = {
+                assert_eq!(mac_bytes.len(), aead::TAG_LEN);
+                let mut buf = [0u8; aead::TAG_LEN];
+                buf.clone_from_slice(&mac_bytes);
+                buf
+            };
+
+            // Test single-shot operation
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: NIT: Comments that are sentences, like this one, should end with a period.

File: src/aead/chacha20_poly1305.rs [hunk 14]
```diff
@@ -256,29 +244,111 @@ extern {
...
+            let key_bytes = test_case.consume_bytes("Key");
+            let input = test_case.consume_bytes("Input");
+            let mac_bytes = test_case.consume_bytes("MAC");
+
+            let key = {
+                assert_eq!(key_bytes.len(), POLY1305_KEY_LEN);
+                let mut buf = [0u8; POLY1305_KEY_LEN];
+                buf.clone_from_slice(&key_bytes);
+                buf
+            };
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: You should be able to use `slice_as_array_ref!` here instead.

File: src/aead/chacha20_poly1305.rs [hunk 15]
```diff
@@ -256,29 +244,111 @@ extern {
...
+
+            let mac = {
+                assert_eq!(mac_bytes.len(), aead::TAG_LEN);
+                let mut buf = [0u8; aead::TAG_LEN];
+                buf.clone_from_slice(&mac_bytes);
+                buf
+            };
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: Ditto: `slice_as_array_ref!`.

File: src/aead/chacha20_poly1305.rs [hunk 16]
```diff
@@ -256,29 +244,111 @@ extern {
...
+
+            // Test single-shot operation
+            let mut state = [0u8; POLY1305_STATE_LEN];
+            let mut mac_out = [0u8; aead::TAG_LEN];
+            poly1305_init(&mut state, &key);
+            poly1305_update(&mut state, &input);
+            poly1305_finish(&mut state, &mut mac_out);
+            assert_eq!(mac, mac_out);
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: Again, please use the `actual_` and `expected_` convention.

File: src/aead/chacha20_poly1305.rs [hunk 17]
```diff
@@ -256,29 +244,111 @@ extern {
...
+
+            // Test streaming byte-by-byte
```
Comment (briansmith) [inline src/aead/chacha20_poly1305.rs]: Ditto period at end of sentence.

**Comment by briansmith** [comment]
Thanks for doing this! Please use the [oxidize-poly1305-tests branch](https://github.com/briansmith/ring/tree/oxidize-poly1305-tests) when you update the PR. Thanks!

File: src/aead/aes_gcm.rs (hunk 5 above)
Comment (newpavlov) [inline src/aead/aes_gcm.rs]: I've renamed it into `output_buf`, as it will be used later to test in-place encryption.

**Comment by newpavlov** [comment]
Thank you for review!

I've merged this PR with `oxidize-poly1305-tests` branch and as far I can see fixed all issues you noted.

**Comment by briansmith** [comment]
@newpavlov I think you are using rustfmt to reformat things but you're not using the rustfmt.toml from this repo. In the future, please use the rustfmt.toml in the repo so that I don't have to re-re-format your changes.

**Comment by briansmith** [comment]
> Thank you for review!
> 
> I've merged this PR with oxidize-poly1305-tests branch and as far I can see fixed all issues you noted.

Unfortunately I cannot figure out a way to review this in a reasonable time. Could you please *rebase* your changes *on top of* the oxidize-poly1305-tests branch. In general, the *ring* project doesn't do merges, only rebases.

Thanks!

**Comment by briansmith** [comment]
OK, I have rebased (I think!) the changes here: https://github.com/briansmith/ring/commits/new-tests. I'll review them shortly.

**Comment by briansmith** [comment]
It turns out that I had missed all the changes that were made in the merge commit. I'm working on merging them back in now.

**Comment by briansmith** [comment]
Landed as 3b5ae496b97ac3414ea51d67b5bd3e6014ffc905 and 6dac94bac6f0dbf69fc9bb69c1ce5a9310d87a48. I tweaked the patches a bit.

Thanks very much for doing this!

**Comment by newpavlov** [comment]
Sorry I couldn't find time to rebase an cleanup it myself.

**Comment by briansmith** [comment]
> Sorry I couldn't find time to rebase and cleanup it myself.

No apologies needed. Everybody is busy!

If you have some time, it might be a good idea to expand the tests in https://github.com/dstucrypt/ring/commits/dstu and maybe convert them over to the data-driven `tests::from_file()` form.
